File: packman/src/main/java/fkn/dlaskina/packman/panels/ConfigPanel.java
package fkn.dlaskina.packman.panels;

import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;

import javax.swing.BoxLayout;
import javax.swing.JButton;
import javax.swing.JComboBox;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JSplitPane;
import javax.swing.JTextArea;
import javax.swing.JTextField;
import javax.swing.text.DefaultCaret;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.Timer;
import java.util.TimerTask;

import static fkn.dlaskina.util.Util.createImageButton;
import static java.awt.GridBagConstraints.BOTH;
import static java.awt.GridBagConstraints.CENTER;

/**
 * Панель конфигурации.
 * @author VLaskin
 * @version 1.0 Date: 13.01.2005 16:38:05
 */
public class ConfigPanel extends JPanel {

    private static final Logger LOG = LogManager.getLogger(ConfigPanel.class);

    private static final Integer[] GEARS = {1, 2, 3, 4, 5};
    /** ссылка на MainFrame. */
    private MainFrame mainFrame = null;
    /** дерево конфигурации. */
    private final JPanel controlPanel = new JPanel(new BorderLayout());
    /** панель для информации об ячейках. */
    private final JTextArea arealInfo = new JTextArea();
    /** поле для ввода команды. */
    private final JTextField commandInput = new JTextField();
    /** поле для ввода расстояния. */
    private final JTextField distanceInput = new JTextField();
    /** поле для ввода угла поворота. */
    private final JTextField angleInput = new JTextField();
    /** поле для ввода передачи. */
    private final JComboBox<Integer> gearComboBox = new JComboBox<>(GEARS);

    /** Обработка события при смене текстовой команды (нажали ENTER в input). */
    private final ActionListener enterOnInputAction = event -> {
        LOG.debug(event.getActionCommand());
        commandInput.setText("");
    };

    public void setMainFrame(final MainFrame mainFrame) {
        this.mainFrame = mainFrame;
    }

    enum CommandActionType {AsIs, Move, Turn};

    /** Обработка события нажатия кнопочки. */
    class CommandActionListener implements ActionListener {

        private final String command;
        private final CommandActionType type;

        public CommandActionListener(final String command, final CommandActionType type) {
            this.command = command;
            this.type = type;
        }

        @Override
        public void actionPerformed(final ActionEvent event) {
            final String oldCom = commandInput.getText();
            String outText = oldCom + command + "30" + ";";
            commandInput.setText(outText);
            LOG.debug(outText);
        }
    };

    /**
     * Конструктор.
     */
    public ConfigPanel() {
        super();

        controlPanel.setBackground(MapPanel.PANEL_GRAY_COLOR);
        fillNavigationPanel();
        fillParametersPanel();

        /** панель для получении информации от робота. */
        arealInfo.setFont(new Font("Arial", Font.PLAIN, 12));
        final JScrollPane plantInfoScroll = new JScrollPane(arealInfo);
        plantInfoScroll.setViewportView(arealInfo);
        ((DefaultCaret) arealInfo.getCaret()).setUpdatePolicy(DefaultCaret.OUT_BOTTOM);

        /* разделительная панелька */
        final JSplitPane splitPane = new JSplitPane(JSplitPane.VERTICAL_SPLIT);
        splitPane.setContinuousLayout(true);
        splitPane.add(controlPanel, JSplitPane.TOP);
        splitPane.add(plantInfoScroll, JSplitPane.BOTTOM);
        //splitPane.setDividerLocation(100);
        //splitPane.setLastDividerLocation(100);
        splitPane.setResizeWeight(0.0);

        setLayout(new BorderLayout());
        /* панель режима. */
        add(splitPane, BorderLayout.CENTER);
    }

    /** создание навигационных кнопок. */
    private void fillNavigationPanel() {
        final JPanel navigationPanel = new JPanel(new GridLayout(3, 4, 5, 5));
        //keyPanel.setSize(150, 150);
        navigationPanel.add(
            createNavigationButton("arrow_up_left.gif", "поворот вперёд и влево", "fl", CommandActionType.Turn)
        );
        navigationPanel.add(
            createNavigationButton("arrow_up.gif", "вперед", "f", CommandActionType.Move)
        );
        navigationPanel.add(
            createNavigationButton("arrow_up_right.gif", "поворот вперёд и вправо", "fr", CommandActionType.Turn)
        );
        navigationPanel.add(
            new JLabel("TL", JLabel.CENTER));
        navigationPanel.add(
            createNavigationButton("arrow_left.gif", "разворот влево наместе", "l", CommandActionType.Turn)
        );
        navigationPanel.add(new JLabel("SC", JLabel.CENTER));
        navigationPanel.add(
            createNavigationButton("arrow_right.gif", "разворот вправо наместе", "r", CommandActionType.Turn)
        );
        navigationPanel.add(new JLabel("TR", JLabel.CENTER));
        navigationPanel.add(
            createNavigationButton("arrow_down_left.gif", "поворот назад и влево", "bl", CommandActionType.Turn)
        );
        navigationPanel.add(
            createNavigationButton("arrow_down.gif", "назад", "b", CommandActionType.Move)
        );
        navigationPanel.add(
            createNavigationButton("arrow_down_right.gif", "поворот назад и вправо", "br", CommandActionType.Turn)
        );
        navigationPanel.add(new JLabel("AUTO", JLabel.CENTER));

        controlPanel.add(navigationPanel, BorderLayout.EAST);

        commandInput.addActionListener(enterOnInputAction);
        controlPanel.add(commandInput, BorderLayout.SOUTH);
    }

    /** создание доп атрибутов. */
    private void fillParametersPanel() {
        distanceInput.setText("30");
        angleInput.setText("90");
        gearComboBox.setSelectedItem(4);
        final JPanel parametersPanel = new JPanel(new GridBagLayout());
        final GridBagConstraints labelGbc = new GridBagConstraints(0, 0, 1, 1, 0, 0,
            CENTER, BOTH, new Insets(2, 5, 2, 4), 0, 0);
        final GridBagConstraints inputGbc = new GridBagConstraints(1, 0, 1, 1, 0, 0,
            CENTER, BOTH, new Insets(2, 0, 2, 0), 0, 0);
        final GridBagConstraints dimensionGbc = new GridBagConstraints(2, 0, 1, 1, 0, 0,
            CENTER, BOTH, new Insets(2, 4, 2, 5), 0, 0);

        parametersPanel.add(new JLabel("Дистанция", JLabel.RIGHT), labelGbc);
        parametersPanel.add(distanceInput, inputGbc);
        parametersPanel.add(new JLabel("[см]", JLabel.LEFT), dimensionGbc);

        labelGbc.gridy++;
        inputGbc.gridy++;
        dimensionGbc.gridy++;
        parametersPanel.add(new JLabel("Угол", JLabel.RIGHT), labelGbc);
        parametersPanel.add(angleInput, inputGbc);
        parametersPanel.add(new JLabel("[град]", JLabel.LEFT), dimensionGbc);

        labelGbc.gridy++;
        inputGbc.gridy++;
        dimensionGbc.gridy++;
        parametersPanel.add(new JLabel("Передача", JLabel.RIGHT), labelGbc);
        parametersPanel.add(gearComboBox, inputGbc);
        parametersPanel.add(new JLabel("gear", JLabel.LEFT), dimensionGbc);

        controlPanel.add(parametersPanel, BorderLayout.WEST);
    }

    private JButton createNavigationButton(
        final String iconName, final String toolTip, final String command, final CommandActionType type
    ) {
        final JButton button = createImageButton(
            null, iconName, 24, 24, toolTip, new CommandActionListener(command, type)
        );
        button.setBorderPainted(true);
        button.setFocusPainted(false);
        button.setContentAreaFilled(true);
        return button;
    }

    public JTextArea getArealInfo() {
        return arealInfo;
    }
}

File: packman/src/main/java/fkn/dlaskina/packman/panels/ConfigPanel.kt
package fkn.dlaskina.packman.panels

import mu.KotlinLogging
import java.awt.BorderLayout
import java.awt.Color
import java.awt.Font
import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import java.awt.Insets
import javax.swing.JButton
import javax.swing.JLabel
import javax.swing.JPanel

private val log = KotlinLogging.logger{}

/**
 * Панель конфигурации.
 * @author VLaskin
 * @version 1.0 Date: 13.01.2005 16:38:05
 */
class ConfigPanel : JPanel() {
    /** всего бонусов  */
    private val lblBonus = JLabel("0", JLabel.LEFT)
    /** собранных бонусов  */
    private val lblBonusMax = JLabel("0", JLabel.LEFT)

    companion object {
        var configPanel: ConfigPanel? = null
        /** число полученых подарков.  */
        private var bonusCount = 0
        /** общее число подарков.  */
        private var bonusCountMax = 0

        /**
         * Добавить подарок и проверить - не последний ли?
         * @return true если это последний бонус
         */
        fun addBonus(): Boolean {
            bonusCount++
            configPanel?.let {
                it.lblBonus.text = bonusCount.toString()
            }
            return bonusCountMax <= bonusCount
        }

        fun setBonusCountMax(maxCount: Int) {
            bonusCount = 0
            bonusCountMax = maxCount
            configPanel?.let {
                it.lblBonusMax.text = bonusCountMax.toString()
            }
        }
    }

    /**
     * Конструктор.
     * final String[] fontNames = GraphicsEnvironment.getLocalGraphicsEnvironment().getAvailableFontFamilyNames();
     * log.info(fontNames);
     */
    init {
        val controlPanel = JPanel(GridBagLayout())
        val c = GridBagConstraints(0, 0, 1, 1, 0.0, 0.0,
                GridBagConstraints.CENTER, GridBagConstraints.BOTH, Insets(2, 10, 2, 10), 0, 0)
        val mainInsets = Insets(2, 10, 2, 10)
        val deskInsets = Insets(2, 10, 2, 2)
        val digitInsets = Insets(2, 5, 2, 10)
        val mainFont = Font("Arial", Font.BOLD, 14)
        val deskFont = Font("Arial", Font.PLAIN, 12)
        val digitFont = Font("Courier New", Font.BOLD, 12)
        c.gridwidth = 2
        c.insets = mainInsets
        var label = JLabel("Призы:", JLabel.CENTER)
        label.font = mainFont
        label.foreground = Color.BLUE
        controlPanel.add(label, c)
        c.gridy = 1
        c.gridx = 0
        c.gridwidth = 1
        c.insets = deskInsets
        label = JLabel("всего:", JLabel.RIGHT)
        label.font = deskFont
        controlPanel.add(label, c)
        c.gridx = 1
        c.insets = digitInsets
        lblBonusMax.font = digitFont
        lblBonusMax.foreground = Color.RED
        controlPanel.add(lblBonusMax, c)
        c.gridx = 0
        c.gridy = 2
        c.insets = deskInsets
        label = JLabel("получено:", JLabel.RIGHT)
        label.font = deskFont
        controlPanel.add(label, c)
        c.gridx = 1
        c.insets = digitInsets
        lblBonus.font = digitFont
        lblBonus.foreground = Color.RED
        controlPanel.add(lblBonus, c)
        c.gridx = 0
        c.gridy = 3
        c.gridwidth = 2
        val button = JButton("Выход")
        controlPanel.add(button, c)
        button.addActionListener { System.exit(0) }
        layout = BorderLayout()
        add(controlPanel, BorderLayout.NORTH)
        configPanel = this
    }
}